### PR TITLE
Make the concurrent-migrator test assert final schema state instead of no-throw

### LINKED ISSUE #1357: Flaky: SerializedMigratorTests.Concurrent_migrators_apply_schema_exactly_once (windows API Integration)
**Classification:** low-frequency environmental flake — investigated, not dismissed (never-merge-failing-CI / never-dismiss-as-flaky discipline).

**Symptom:** `SerializedMigratorTests.Concurrent_migrators_apply_schema_exactly_once` intermittently fails on `API Integration (windows-latest)` with `Microsoft.Data.Sqlite.SqliteException: SQLite Error 1: 'table "ProposalProvenances" already exists'` — i.e. two concurrent migrators both applied an *old* migration, so the advisory lock did not serialize them that run.

**Evidence it is flaky (not a regression):** failed once on `main@04e7969d` (the 2026-07-13 revival/GEN wave head), then **passed on `main@1f65fc24`** (docs-only diff — identical `SerializedMigrator` + migration code). Not in the failure ledger before; not the known #1282/#1335 hosted-worker race.

**Root-cause read:** the test runs 2 migrators against a fresh cold DB behind a `Barrier`, 4 iterations. The wave added 2 migrations (`AddRegistrationGating`, `AddArtefactExtractions`), lengthening the cold-DB chain and widening the window where both migrators can take the lock-free fail-open path (uncreatable/timed-out lock under CI load) and collide mid-chain; the loser's `NoMigrationsPending` re-check sees migrations still pending (winner not finished) and rethrows. `SerializedMigrator` itself (WAL + busy_timeout + advisory file lock + pending re-check) is sound; this is test-harness concurrency fragility on shared CI.

**Suggested hardening (pick one):**
- Make the assertion state-based: allow the documented fail-open collision but assert the *final* schema is applied exactly once (`__EFMigrationsHistory` complete, no dupes), rather than `NotThrowAsync`.
- Or give the test a longer lock timeout and/or fewer iterations to reduce the CI-load window.
- Or investigate why the sidecar lock file becomes uncreatable/contended under Windows CI load and make `AcquireLock` retry that case.

**Severity:** LOW (test-only; product mechanism is sound). Reruns green.

### PR DESCRIPTION
**Pull request by Chris0Jeky**
Closes #1357

## Problem

`SerializedMigratorTests.Concurrent_migrators_apply_schema_exactly_once` intermittently failed on `API Integration (windows-latest)` with `SQLite Error 1: 'table "ProposalProvenances" already exists'`. The test asserted `NotThrowAsync` — but that assertion is wrong for what `SerializedMigrator` actually guarantees.

## Root cause (verified against the code)

`SerializedMigrator` has a **documented lock-free fail-open path** (`AcquireLock` returns `null` when the sidecar `.migrate.lock` file is uncreatable or its acquisition times out; `Migrate` then calls `MigrateWithoutLock`). On that path a losing migrator races the winner mid-chain. EF computes the pending set **before** taking SQLite's write lock, so the loser re-issues DDL the winner already applied and throws (or hits a duplicate `__EFMigrationsHistory` insert). `MigrateWithoutLock` swallows that throw **only** when a `NoMigrationsPending` re-check shows nothing remains pending; if the winner has not finished the chain, the re-check still sees pending migrations and **rethrows**.

The 2026-07-13 wave added `AddRegistrationGating` + `AddArtefactExtractions`, lengthening the cold-DB chain (37 migrations) and widening the window where, under Windows CI load, both migrators take the fail-open path and collide mid-chain before the winner finishes. This is **test-harness fragility asserting the wrong thing**, not a product defect — `SerializedMigrator` (WAL + busy_timeout + advisory file lock + pending re-check) is sound, and the fail-open collision is explicitly documented behavior.

## Fix

State-based assertions instead of no-throw, plus adversarial-review hardening (two independent lenses, coordinator-adjudicated — see review comments):

1. **Concurrent test asserts final schema state.** Await both migrators; collect faults. Tolerate **only** the SQLite collision signature — **any** `SqliteException`, walked through the inner-exception chain. Message-matching was rejected in review (HIGH): the same race lands on whatever DDL the colliding migration issues first — `already exists` (CreateTable), `duplicate column name` (AddColumn-first migrations like `AddDeferredUntilToAutomationProposal`), `no such table`/`ef_temp` (table rebuilds), `UNIQUE constraint failed` (history insert) — so string enumeration re-flakes on the next chain reshape. The type check only excludes non-SQL faults (barrier timeouts, null refs, IO), which stay hard failures; the real gates are `faults.Count < 2` (the winner never throws the collision it caused, so both faulting means no run applied the chain) plus the exact-state block: `__EFMigrationsHistory` complete (`BeEquivalentTo(GetMigrations())`), duplicate-free (`OnlyHaveUniqueItems`), and a **usable** representative table (`verify.Set<Board>().Count()` runs a real query). Theory coverage pins the signature helper for every collision shape, a wrapped `DbUpdateException`, and non-SQL rejection.
2. **Deterministic fail-open collision test.** Under normal conditions the loser acquires the lock and no-ops, so green runs prove the lock, not the tolerance. New test holds the sidecar lock externally (same technique as the existing timeout test) with a 250ms lock timeout so BOTH migrators fail open, barrier-syncs them on a cold DB so both compute the full pending chain, and races them: at most 1 fault, SQLite collision signature on any fault, both fail-open warnings asserted, exactly-once final state, and collision evidence (propagated fault or the swallowed-race warning) required within 3 fresh-DB attempts.
3. **AcquireLock exclusivity contract test.** Removing the no-throw assertion deleted the suite's only lock-serialization regression guard — a `FileShare.None → ReadWrite` weakening (or early stream disposal) would have passed every state-based test, silently disabling the #1164 contract. `AcquireLock` goes `private → internal` (visibility only, via the pre-existing `InternalsVisibleTo("Taskdeck.Api.Tests")`), and a direct contract test asserts: held lock excludes a second acquisition (fail-open `null` + warning) and is reacquirable after release.

Files: `backend/tests/Taskdeck.Api.Tests/SerializedMigratorTests.cs`, `backend/src/Taskdeck.Infrastructure/Persistence/SerializedMigrator.cs` (visibility-only, no behavior change).

## What this does and does not claim

The CI flake is **tolerated, not eliminated**: under CI load the advisory lock legitimately degrades to the documented fail-open path, and the test now accepts that outcome while pinning the invariant that matters (final schema migrated exactly once, usable, complete history). The fail-open collision itself is now deterministically covered (fix 2), so the tolerance branch is proven, not just permitted. The 10-second barrier wait is a deliberate fail-fast posture: if a sibling migrator dies before the barrier, the test fails fast instead of hanging CI.

## Verification

- `dotnet restore` once; `dotnet build backend/Taskdeck.sln -c Release -m:1` → **0 errors**.
- `--filter FullyQualifiedName~SerializedMigratorTests` (now 12 tests: 4 original-shape + 6 signature theory cases + fail-open collision + exclusivity contract) run **10 consecutive times → 10/10 green** (evidence in the fix-evidence comment).

## Discrimination proof (state assertion is not false-green)

Temporarily tampered the shipped equivalence assertion to expect one migration the DB does not contain:
`historyIds.Should().BeEquivalentTo(verify.Database.GetMigrations().Append("0000_DISCRIMINATION_PROBE"))`.
The test went **RED**: `Expected historyIds to be a collection with 38 item(s) ... but {...37 real ids...} contains 1 item(s) less than`. This confirms the assertion reads the real migrated history from the DB and fails when the final schema is genuinely wrong. The probe was reverted and **not committed**.

## Adversarial review

Two independent adversarial lenses were run and coordinator-adjudicated: 1 HIGH + 2 MEDIUM + 1 LOW confirmed and fixed in this batch (matcher gap, untested tolerance branch, deleted serialization guard, PR-body wording); 1 finding refuted with evidence (the failure site is this test itself per issue #1357's recorded symptom — "windows API Integration" is the CI job that runs `Taskdeck.Api.Tests`, not a different code path). Full findings and fix-evidence mapping are in the PR comments.

- **Can the tolerance mask a real product regression?** No. If the lock were completely broken and both migrators applied the chain, SQLite still serializes writes — one wins per migration, the loser throws the tolerated collision, and the final-state block still demands a single complete history and usable schema (proven red-able by the discrimination probe). A lock that silently stops excluding is now caught directly by the exclusivity contract test rather than indirectly by no-throw.
- **Locked-path regressions still surface**: the locked `Migrate()` never swallows, so a genuine failure there fails the test regardless of signature tolerance.

## Possible follow-up (deferred, not in this PR)

The issue also suggested making `AcquireLock` retry the uncreatable-lock-file case under Windows contention. `AcquireLock` currently fail-opens immediately on `UnauthorizedAccessException` (which Windows raises transiently for a lock file pending deletion). A bounded retry would reduce fail-open frequency, but it changes production concurrency behavior, is not deterministically testable, and — done naively — would make a genuinely unwritable directory spin for the full 30s timeout before failing open (a startup-stall regression) unless transient vs. permanent `ACCESS_DENIED` are distinguished. Not cheap/clean enough for this scope; the fail-open path is now deterministically covered by tests either way. Recommend a separate tracked issue if the fail-open frequency is worth reducing.

### CODE REVIEW

**Comment by Chris0Jeky** [comment]
## Adversarial Code Review

Self-review of the state-based hardening. Scope is test-only (`SerializedMigratorTests.cs`); no production `SerializedMigrator` change.

### CRITICAL
- None.

### HIGH
- None. **Masking analysis (cleared):** exception-tolerance is signature-based, not path-based, so a broken *locked* path that threw `already exists` would be tolerated at the exception level. But tolerance never gates success on its own — the definitive guard is the final-state block. For the schema to end up wrong you need a missing migration, a duplicate history row, or an unusable table, and each is asserted (`BeEquivalentTo(GetMigrations())`, `OnlyHaveUniqueItems`, `Set<Board>().Count()`). "Both migrators applied the chain twice" cannot complete (SQLite serializes writes; the `MigrationId` PK and `CREATE TABLE` reject the second application → the loser throws, the winner completes), so there is no false-green path. Confirmed empirically by the discrimination probe (see PR body).

### MEDIUM
- None. **Empty-`GetMigrations()` false-green (cleared):** `BeEquivalentTo(GetMigrations())` alone could green an all-empty run if `GetMigrations()` ever returned empty. It does not (37 compiled migrations), and the `Boards` existence + usability checks fail on an empty schema regardless — defense in depth is already present.

### LOW
- **L1 — Redundant `Boards` checks.** `GetUserTableNames(verify).Should().Contain("Boards")` is subsumed by `verify.Set<Board>().Count()` (if the query works, the table exists). **Explicitly invalidated (kept intentionally):** the two serve distinct diagnostics — the name check pinpoints "table never created", the query check pinpoints "table created but malformed/unusable". Collapsing them would blur failure attribution. No change.
- **L2 — `SQLITE_BUSY` deliberately excluded from tolerance.** `IsDocumentedMidChainCollision` tolerates only `already exists` / `UNIQUE constraint failed`, not `database is locked`. If `busy_timeout` (5000ms in-test) were ever exceeded under CI load, a leaked `SQLITE_BUSY` would fail the hardened test rather than being absorbed. **Intentional, not a fix:** the observed flake (#1357) is the logical DDL collision (Error 1), and `SerializedMigrator`'s contract states `busy_timeout` serializes `SQLITE_BUSY`, so a leak there is a genuinely different defect that must surface, not be masked. Noted as residual risk.

### Bot Comments Addressed
- None present yet (no human, review, or bot comments on the PR; Codex review posts on push and may arrive later — will address if it does).

### Summary
0 CRITICAL, 0 HIGH, 0 MEDIUM, 2 LOW (both explicitly invalidated/intentional with reasoning). No code changes required. Not merge-blocking. Evidence: Release build 0 errors; `--filter FullyQualifiedName~SerializedMigratorTests` 10/10 green; discrimination probe proved the state assertion goes red on a wrong final schema (reverted, not committed).

**Comment by Chris0Jeky** [comment]
## Adversarial Review — Fixes Applied

| Finding | Severity | Disposition | Verified |
|---------|----------|-------------|----------|
| Locked-path throw tolerated at exception level | HIGH (cleared) | Not a defect — final-state block is the real guard; no false-green path exists | Discrimination probe (red on wrong schema) |
| Empty `GetMigrations()` false-green | MEDIUM (cleared) | Not reachable — 37 compiled migrations + `Boards` existence/usability checks | 10/10 green; probe |
| L1 Redundant `Boards` name vs usability check | LOW | Explicitly invalidated — kept for distinct diagnostics | n/a |
| L2 `SQLITE_BUSY` excluded from tolerance | LOW | Intentional per documented signature — residual risk noted | n/a |

No code changes were required: zero CRITICAL/HIGH/MEDIUM defects, and both LOW findings are explicitly invalidated/intentional with reasoning. Nothing to push. All findings addressed; none silently dropped.

CI status at review time: **PENDING** (fresh run). The change is test-only and the hardened class is 10/10 green locally; will not merge (worktree worker never merges).

**Comment by Chris0Jeky** [comment]
## Consolidated Adversarial Review — Two Independent Lenses (coordinator-adjudicated)

Two independent adversarial reviews of this PR were run and adjudicated. The state-based direction is **confirmed sound** (double-fault impossibility, history-equivalence catching mid-chain death, and the `SQLITE_BUSY` mitigation via the 5000ms busy_timeout interceptor all verified). Four findings survive adjudication and will be fixed in this batch; one finding was refuted.

### HIGH

- **F1 (flake lens, CONFIRMED) — matcher gap in `IsDocumentedMidChainCollision`.** The helper only matches `already exists` / `UNIQUE constraint failed`, but the *same* fail-open race landing on an AddColumn migration throws `duplicate column name: X` (the chain has AddColumn-first migrations, e.g. `20260626234137_AddDeferredUntilToAutomationProposal`), and table-rebuild migrations (AddForeignKey/DropColumn) can race into `no such table` / `ef_temp` shapes. The test would re-flake with a misleading "real defect" message on the next unlucky collision site. **Fix:** stop enumerating message strings — tolerate ANY `SqliteException` (walked through the inner-exception chain, any error code). This cannot mask a real defect because the exception-type check never gates success alone: the `faults.Count < 2` gate rejects "no migrator drove the chain" (a genuinely broken migration faults BOTH migrators and also fails the single-migrator test), and the exact-state assertions (complete duplicate-free history == `GetMigrations()`, usable schema) reject any incomplete/corrupted outcome. The type check exists only to keep non-SQL faults (timeouts, null refs, IO, barrier failures) as hard failures. Plus unit-fixture coverage of the helper with constructed exceptions.

### MEDIUM

- **F2 (flake lens, CONFIRMED) — tolerance branch never exercised.** Under normal conditions the loser acquires the lock and no-ops, so 10/10 green proves the lock, not the tolerance. **Fix:** add a deterministic fail-open collision test — hold the sidecar lock file externally (same technique as `Migrate_proceeds_with_a_warning_when_lock_cannot_be_acquired_within_timeout`) with a short lock timeout so BOTH migrators fail open, barrier-sync them on a cold DB so both compute the full pending chain, then race. Assert: at most 1 fault, any fault matches the collision signature, final state exactly-once (same helpers), and both migrators actually logged the fail-open warning.
- **F3 (masking lens, adjudicated real) — deleted the suite's only lock-serialization regression guard.** Removing the no-throw assertion means a `FileShare.None → ReadWrite` regression in `SerializedMigrator.AcquireLock` (or early lock disposal) would now pass everything, silently disabling the #1164 serialization contract. **Fix:** add a deterministic lock-exclusivity contract test at the `AcquireLock` seam — while one acquisition holds, a second acquisition of the same path must fail open (null) within a short timeout; after release, acquisition must succeed. `AcquireLock` goes `private → internal`; `Taskdeck.Infrastructure` already declares `InternalsVisibleTo("Taskdeck.Api.Tests")`, so no new visibility arrangement is introduced.

### LOW

- **F4 (flake lens F2/F5) — PR-body honesty.** The body claims the test hardening "fully resolves the flake". Reword: the flake is **tolerated** where the advisory lock legitimately degrades under CI load (and is now deterministically covered via F2's test), not eliminated. Also note the 10s barrier fail-fast is a deliberate posture (hang becomes a fast fail).

### Refuted (no action)

- **Flake lens Finding 3** claimed the original #1357 failure site was the TestWebApplicationFactory raw-`Database.Migrate()` path. Wrong premise: issue #1357's recorded symptom names the test itself — *"`SerializedMigratorTests.Concurrent_migrators_apply_schema_exactly_once` intermittently fails on `API Integration (windows-latest)` with `SQLite Error 1: 'table "ProposalProvenances" already exists'`"*. "windows API Integration" is the CI **job** that runs `Taskdeck.Api.Tests`, not a different code path.

### Summary

1 HIGH, 2 MEDIUM, 1 LOW confirmed — all four being fixed in this batch (no dismissals). 1 finding refuted with evidence. Fix-evidence comment with commit mapping and fresh 10x verification to follow.

**Comment by Chris0Jeky** [comment]
## Adversarial Review — Fixes Applied (consolidated batch)

| Finding | Severity | Fix Commit | Verified |
|---------|----------|-----------|----------|
| F1 — matcher gap: message-string enumeration re-flakes on AddColumn (`duplicate column name`) / table-rebuild (`no such table`, `ef_temp`) collision shapes | HIGH | `bda90556` | Signature helper now tolerates ANY `SqliteException` walked through the inner-exception chain; 6 theory cases pin every collision shape + wrapped `DbUpdateException` + non-SQL rejection. Class green. |
| F2 — tolerance branch never exercised (green runs proved the lock, not the tolerance) | MEDIUM | `ca6023f9` | New `Fail_open_collision_is_tolerated_and_schema_still_applies_exactly_once`: sidecar lock held externally + 250ms timeout forces BOTH migrators onto the fail-open path (warnings asserted), barrier-synced cold-DB race, ≤1 fault, SQLite signature on any fault, exactly-once final state, collision evidence required within 3 attempts. Green 10/10. |
| F3 — deleted the suite's only lock-serialization regression guard (#1164) | MEDIUM | `c904b9f2` + `fea23c24` | `AcquireLock` `private → internal` (visibility only, via pre-existing `InternalsVisibleTo("Taskdeck.Api.Tests")`); new `AcquireLock_is_exclusive_while_held_and_reacquirable_after_release` asserts: held lock excludes a second acquisition (fail-open `null` + documented warning), reacquirable after release. A `FileShare.None` weakening or early disposal now fails deterministically. |
| F4 — PR-body honesty ("resolves the flake" overclaim; barrier posture undocumented) | LOW | PR body edited (no code) | Body now states the flake is **tolerated** where the advisory lock legitimately degrades under CI load (deterministically covered via F2), not eliminated; 10s barrier fail-fast documented as deliberate (hang → fast fail). `Closes #1357` link re-verified intact after the edit. |
| Flake-lens Finding 3 (TestWebApplicationFactory raw-`Migrate()` as the failure site) | — | REFUTED, no action | Issue #1357's recorded symptom names `SerializedMigratorTests.Concurrent_migrators_apply_schema_exactly_once` itself; "windows API Integration" is the CI job that runs `Taskdeck.Api.Tests`, not a different code path. |

### Verification evidence (head `fea23c24`)

- Full-solution `dotnet build -c Release -m:1`: **0 errors**.
- `--filter FullyQualifiedName~SerializedMigratorTests` (12 tests: 4 original-shape + 6 signature theory + fail-open collision + exclusivity contract) — **10/10 consecutive green** runs: 12/12 passed each run (durations 8–21s).
- `Taskdeck.Architecture.Tests`: 20 passed, 0 failed (the `internal` visibility change violates no layer boundary).
- Discrimination probe (earlier, unchanged assertion): tampered expected list → RED (`38 item(s) ... contains 1 item(s) less than`); reverted, not committed.

All confirmed findings fixed; refuted finding answered with evidence; nothing dropped. CI status: PENDING (fresh push). Will not merge (worktree worker never merges).

File: backend/tests/Taskdeck.Api.Tests/SerializedMigratorTests.cs [hunk 1]
```diff
@@ -82,23 +82,60 @@ Task RunMigratorAsync(TaskdeckDbContext context) => Task.Run(() =>
...
+                        faults.Add(ex);
+                    }
+                }
+
+                foreach (var fault in faults)
+                {
+                    IsDocumentedMidChainCollision(fault).Should().BeTrue(
+                        "the only tolerable concurrent-migrator failure is the documented fail-open " +
+                        "mid-chain collision (\"already exists\" / duplicate __EFMigrationsHistory); " +
+                        $"any other throw is a real defect, but got: {fault}");
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/tests/Taskdeck.Api.Tests/SerializedMigratorTests.cs]: Confirmed — this is finding F3 in the consolidated adversarial review (adjudicated real) and is fixed as of head fea23c24, pushed after the commit you reviewed (19284e1d):

1. **Direct exclusivity contract test** (c904b9f2 + fea23c24): `AcquireLock` is now `internal` (visibility only, via the pre-existing `InternalsVisibleTo("Taskdeck.Api.Tests")`), and `AcquireLock_is_exclusive_while_held_and_reacquirable_after_release` asserts that a held lock excludes a second acquisition of the same path (fail-open `null` + documented warning) and is reacquirable after release. A `FileShare.None` weakening or early stream disposal — the broken-lock scenario you describe — now fails this test deterministically instead of being masked by the state-based assertions.

2. **Fail-open path is now proven, not inferred** (ca6023f9): the new deterministic collision test passes per-migrator loggers and asserts the `could not acquire migration lock` warning on BOTH migrators, so the tolerance branch is exercised on the actual fail-open path, distinguishable from the locked path exactly as you note.

The pre-existing `Single_migrator_on_file_db_acquires_lock_and_does_not_fail_open` test additionally keeps asserting the locked-path Debug entry with zero warnings on the happy path. Evidence: full class (12 tests) 10/10 consecutive green at fea23c24.
